### PR TITLE
fix(ubuntu) correct asdf inclusion in `/etc/environment` for Linux VMs

### PR DESCRIPTION
**Pull request by lemeurherve**
This PR uses hardcoded path to include asdf in the $PATH so asdf installed tools like yq can be executed.

Note: there might be a way to keep this folder as a variable but as it breaks current uses this "easy" solution should be enough, can be revised in a following pull request.

Ref:
- #1042

### CODE REVIEW

**Comment by lemeurherve** [comment]
On arm images `goss` seems to pick the `ruby` version installed by `vagrant` as one of its dependencies (3.0.2p107), while `asdf` is installing the version 2.6.10 **after** `vagrant` 🤔

(Search for "ruby" in https://infra.ci.jenkins.io/job/infra-tools/job/packer-images/job/PR-1043/2/pipeline-console/log?nodeId=120)

**Review by timja — APPROVED** [review]

**Review by smerle33 — CHANGES_REQUESTED** [review]
this doesn't correct the problem as seen here https://infra.ci.jenkins.io/job/infra-tools/job/packer-images/job/PR-1043/13/ 
but I think, point to why I cannot use the correct npm here https://github.com/jenkins-infra/helpdesk/issues/3823#issuecomment-1918841628

**Comment by dduportal** [comment]
Please note that you will never get a universal solution as the `PATH` is governed by a LOT of different mechanisms (considering only Linux):

- Jenkins environment global setup (if any)
- Jenkins per-agent environment setup (if any) => usually only global agents
- Jenkins agent/cloud plugin environment setup (if any) => different between Azure VM, Kubernetes, Azure Container Instances, EC2, etc.
- Jenkins agent initialization method => inbound (relies on the agent process spawning method) or SSH (relies on the SSH server setup which **may** filter the `PATH`)
- Jenkins agent type => major differences between VMs (systemd load different environments) and containers (no `/etc/environment` loaded but image metadata are used to define environment)
- Shell used for the pipeline (global `bash`, local `bash`) => we enforce `/bin/bash` at Jenkins controller but it can be overloaded with the proper pipeline syntax
- Pipeline environment (`env.` groovy object and the `environement {}` /  `withEnv[] {}` blocks)
-

File: build-jenkins-agent-ubuntu.pkr.hcl [hunk 1]
```diff
@@ -51,9 +51,9 @@ build {
   }
 
   provisioner "shell" {
-    execute_command = "{{ .Vars }} sudo -E su - jenkins -c \"bash -eu '{{ .Path }}'\""
+    execute_command = "{{ .Vars }} sudo -E su - jenkins -c \"bash -eux '{{ .Path }}'\""
     inline = [
-      "source /home/jenkins/.asdf/asdf.sh", # Required as this is a non-interactive and non-login `bash`
+      "source /home/jenkins/.asdf/asdf.sh && asdf list", # Required as this is a non-interactive and non-login `bash`
```
Comment (dduportal) [inline build-jenkins-agent-ubuntu.pkr.hcl:56]: Why `asdf list` ? Is it debug?

Comment (lemeurherve) [reply build-jenkins-agent-ubuntu.pkr.hcl:56]: Yes (cf commit msg)

File: provisioning/ubuntu-provision.sh [hunk 2]
```diff
@@ -156,7 +156,7 @@ function install_asdf() {
   rm -f "${archive}"
 
   ## append to the system wide path variable, need to be seconded for docker in packer sources.pkr.hcl
-  sed -e "/^PATH/s/\"$/:\${asdf_install_dir}\/shims:\${asdf_install_dir}\/bin\"/g" -i /etc/environment
+  sed -e '/^PATH/s/"$/:\/home\/jenkins\/.asdf\/shims:\/home\/jenkins\/.asdf\/bin\"/g' -i /etc/environment
```
Comment (lemeurherve) [inline provisioning/ubuntu-provision.sh:159]:
```suggestion
  sed --regexp-extended "s|^PATH=(.*)"$|PATH=\1:${asdf_install_dir}/shims:${asdf_install_dir}/bin"|" -i /etc/environment
```
Result of mob programming with @smerle33 @dduportal

File: build-jenkins-agent-ubuntu.pkr.hcl [hunk 3]
```diff
@@ -51,9 +51,9 @@ build {
...
       "goss --version",
```
Comment (lemeurherve) [inline build-jenkins-agent-ubuntu.pkr.hcl:57]:
```suggestion
      "echo $PATH && cat /etc/environment",
      "goss --version",
```
More debug

File: provisioning/ubuntu-provision.sh [hunk 4]
```diff
@@ -156,7 +156,7 @@ function install_asdf() {
   rm -f "${archive}"
 
   ## append to the system wide path variable, need to be seconded for docker in packer sources.pkr.hcl
-  sed -e "/^PATH/s/\"$/:\${asdf_install_dir}\/shims:\${asdf_install_dir}\/bin\"/g" -i /etc/environment
+  sed --regexp-extended "s|^PATH=(.*)"$|PATH=\1:${asdf_install_dir}/shims:${asdf_install_dir}/bin"|" -i /etc/environment
```
Comment (lemeurherve) [inline provisioning/ubuntu-provision.sh:159]:
```suggestion
  sed --regexp-extended "s|^PATH=(.*)"$|PATH=\1:${asdf_install_dir}/shims:${asdf_install_dir}/bin"|" /etc/environment
```

> [2024-02-12T09:51:41.175Z] [1;31m2024-02-12T09:51:41Z: ==> docker.ubuntu: + sed --regexp-extended 's|^PATH=(.*)$'[0m
> [2024-02-12T09:51:41.175Z] [1;31m2024-02-12T09:51:41Z: ==> docker.ubuntu: + PATH='1:/home/jenkins/.asdf/shims:/home/jenkins/.asdf/bin|'[0m
> [2024-02-12T09:51:41.175Z] [1;31m2024-02-12T09:51:41Z: ==> docker.ubuntu: + -i /etc/environment[0m
> [2024-02-12T09:51:41.175Z] [1;31m2024-02-12T09:51:41Z: ==> docker.ubuntu: /tmp/script_8682.sh: line 159: -i: command not found[0m

**Comment by lemeurherve** [comment]
@smerle33 @dduportal KO:

> 
> [2024-02-12T10:16:16.236Z] [1;31m2024-02-12T10:16:16Z: ==> azure-arm.ubuntu: + sed --regexp-extended 's|^PATH=(.*)$'[0m
> [2024-02-12T10:16:16.236Z] [1;31m2024-02-12T10:16:16Z: ==> azure-arm.ubuntu: + PATH='1:/home/jenkins/.asdf/shims:/home/jenkins/.asdf/bin|'[0m
> [2024-02-12T10:16:16.236Z] [1;31m2024-02-12T10:16:16Z: ==> azure-arm.ubuntu: + /etc/environment[0m
> [2024-02-12T10:16:16.236Z] [1;31m2024-02-12T10:16:16Z: ==> azure-arm.ubuntu: /tmp/script_60.sh: line 159: /etc/environment: Permission denied[0m
> [2024-02-12T10:16:16.236Z] [1;31m2024-02-12T10:16:16Z: ==> azure-arm.ubuntu: sed: -e expression #1, char 13: unterminated `s' command[0m

File: provisioning/ubuntu-provision.sh [hunk 5]
```diff
@@ -156,7 +156,8 @@ function install_asdf() {
   rm -f "${archive}"
 
   ## append to the system wide path variable, need to be seconded for docker in packer sources.pkr.hcl
-  sed -e "/^PATH/s/\"$/:\${asdf_install_dir}\/shims:\${asdf_install_dir}\/bin\"/g" -i /etc/environment
+  ## https://backreference.org/2010/02/20/using-different-delimiters-in-sed/index.html
+  sed -R -E "s|^PATH=(.*)"$|PATH=\1:${asdf_install_dir}/shims:${asdf_install_dir}/bin"|" /etc/environment
```
Comment (lemeurherve) [inline provisioning/ubuntu-provision.sh:160]:
```suggestion
  sed -R "s|^PATH=\(.*\)"$|PATH=\1:${asdf_install_dir}/shims:${asdf_install_dir}/bin"|" /etc/environment
```

File: provisioning/ubuntu-provision.sh [hunk 6]
```diff
@@ -156,7 +156,8 @@ function install_asdf() {
   rm -f "${archive}"
 
   ## append to the system wide path variable, need to be seconded for docker in packer sources.pkr.hcl
-  sed -e "/^PATH/s/\"$/:\${asdf_install_dir}\/shims:\${asdf_install_dir}\/bin\"/g" -i /etc/environment
+  ## https://backreference.org/2010/02/20/using-different-delimiters-in-sed/index.html
+  sed --regexp-extended "s|^PATH=\(.*\)\"$|PATH=\1:${asdf_install_dir}/shims:${asdf_install_dir}/bin\"|" /etc/environment
```
Comment (dduportal) [inline provisioning/ubuntu-provision.sh:160]:
```suggestion
  sed --in-place --regexp-extended "s|^PATH=(.*)\"|PATH=\1:${asdf_install_dir}/shims:${asdf_install_dir}/bin\"|g" /etc/environment
```

Comment (dduportal) [reply provisioning/ubuntu-provision.sh]: - Ref.: https://www.gnu.org/software/sed/manual/html_node/ERE-syntax.html , the `--regexp-extended` (same as `-r` and `-E`) means we must not escape special characters such as parenthesis (`(` / `)`) unless we want them literate. Our CI errors was due to no escape group in the search pattern: as such `\1` was unknown
- The `g` at the end of the sed argument has no influence: I've added it by reflex

File: build-jenkins-agent-ubuntu.pkr.hcl [hunk 7]
```diff
@@ -51,9 +51,10 @@ build {
   }
 
   provisioner "shell" {
-    execute_command = "{{ .Vars }} sudo -E su - jenkins -c \"bash -eu '{{ .Path }}'\""
+    execute_command = "{{ .Vars }} sudo -E su - jenkins -c \"bash -eux '{{ .Path }}'\""
```
Comment (dduportal) [inline build-jenkins-agent-ubuntu.pkr.hcl:54]:
```suggestion
    execute_command = "{{ .Vars }} sudo -E su - jenkins -c \"bash -eu '{{ .Path }}'\""
```

File: build-jenkins-agent-ubuntu.pkr.hcl [hunk 8]
```diff
@@ -51,9 +51,10 @@ build {
...
     inline = [
-      "source /home/jenkins/.asdf/asdf.sh", # Required as this is a non-interactive and non-login `bash`
+      "source /home/jenkins/.asdf/asdf.sh && asdf list", # Required as this is a non-interactive and non-login `bash`
```
Comment (dduportal) [inline build-jenkins-agent-ubuntu.pkr.hcl:56]:
```suggestion
      "source /home/jenkins/.asdf/asdf.sh", # Required as this is a non-interactive and non-login `bash`
```

File: build-jenkins-agent-ubuntu.pkr.hcl [hunk 9]
```diff
@@ -51,9 +51,10 @@ build {
...
+      "echo $PATH && cat /etc/environment",
```
Comment (dduportal) [inline build-jenkins-agent-ubuntu.pkr.hcl:57]:
```suggestion
```

File: provisioning/ubuntu-provision.sh [hunk 10]
```diff
@@ -156,7 +156,8 @@ function install_asdf() {
   rm -f "${archive}"
 
   ## append to the system wide path variable, need to be seconded for docker in packer sources.pkr.hcl
-  sed -e "/^PATH/s/\"$/:\${asdf_install_dir}\/shims:\${asdf_install_dir}\/bin\"/g" -i /etc/environment
+  ## https://backreference.org/2010/02/20/using-different-delimiters-in-sed/index.html
+  sed --in-place --regexp-extended "s|^PATH=(.*)\"|PATH=\1:${asdf_install_dir}/shims:${asdf_install_dir}/bin\"|g" /etc/environment
```
Comment (dduportal) [inline provisioning/ubuntu-provision.sh:160]:
```suggestion
  sed --in-place --regexp-extended "s|^PATH=(.*)\"$|PATH=\1:${asdf_install_dir}/shims:${asdf_install_dir}/bin\"|" /etc/environment
```

**Review by dduportal — CHANGES_REQUESTED** [review]

**Review by smerle33 — APPROVED** [review]

**Review by dduportal — APPROVED** [review]
Approving the PR, as we need to land the `/etc/environment` fix first before dealing with the `PATH` order of the installed ruby (package dependency of `vagrant`).

We'll update the goss test for `ruby` to only check for the presence of a ruby binary in the default PATH, not a specific version

**Review by dduportal — APPROVED** [review]

**Pull request merged**